### ISSUE TITLE
Lane assist

### ISSUE THREAD

**Issue by Donkey-Doug** [organicmaps/organicmaps#1404]
Apple and Google maps both have it. I prefer Apple's implementation:

Apple maps:
![Apple-MAps-navigation-teaser](https://user-images.githubusercontent.com/50382414/136707480-1f101b6a-fc5b-42cc-bc9e-dfa63cfd913b.jpg)

Google maps:
![Screenshot_20211010-194505_Firefox](https://user-images.githubusercontent.com/50382414/136707487-8c11d67d-32a5-4599-b0dc-32d4fbeb49f7.png)

**Comment by biodranik** [comment]
See also #1104 
Lanes are not well-mapped yet in OpenStreetMap. But it's a great idea, of course.

**Comment by Donkey-Doug** [comment]
> Mapping is far from being completed but another app using this data would likely cause increase in the mapping velocity - it's quite easy to map from satellite imagery.

Interesting

**Comment by Donkey-Doug** [comment]
Do we know what OSMAND's data source is?

**Comment by starsep** [comment]
> Do we know what OSMAND's data source is?
> 
> Or can we reverse engineer their implementation?
OsmAnd is open source, no need for reverse engineering :).
Their first implementation is here: https://github.com/osmandapp/OsmAnd/commit/d6f1be4f80a2d7e0d9580200a1b4090729cfa694 (old commit)
It seems to be mostly implemented here https://github.com/osmandapp/OsmAnd/blob/master/OsmAnd-java/src/main/java/net/osmand/router/RouteResultPreparation.java using `turn:lanes`, `turn:lanes:forward`, and `turn:lanes:backward`

**Comment by Donkey-Doug** [comment]
Oh, thats great! I did not know lanes were mapped in OSM.

**Comment by Donkey-Doug** [comment]
Some more implementations:

Sygic:
![lane-assist](https://user-images.githubusercontent.com/50382414/137905456-183b69cb-b6d6-4cca-9988-f0f30ff9e00a.png)

Here wego:

![Lane-assistance_0](https://user-images.githubusercontent.com/50382414/137905514-3c8e302b-84a0-4da5-9557-192c38f774f0.png)

**Comment by zyphlar** [comment]
Here is OsmAnd's lane indicator, for sake of completeness. It's useful although the colors take awhile to get used to (red yellow and green don't really map intuitively to long medium and short distances.)

![](https://i.imgur.com/PXdY4mm.jpeg)

**Comment by suhridkhan** [comment]
> Here is OsmAnd's lane indicator, for sake of completeness. It's useful although the colors take awhile to get used to (red yellow and green don't really map intuitively to long medium and short distances.)
> 
> ![](https://camo.githubusercontent.com/5f1bba98d4fe3a4735e9800dfb846f0657b56b2b85ad723f5d30fc3b80c9e034/68747470733a2f2f692e696d6775722e636f6d2f50586459346d6d2e6a706567)

Implementation just like OsmAnd should be adequate, and quite useful feature.

**Comment by tibordoma** [comment]
Any updates? For me, this is quite important feature. I like also Mapy.cz implementation (they use OSM data):[](url)
![mapy-cz-navigace-a-dopravni-mapa-3](https://user-images.githubusercontent.com/13929686/179606066-5b523ec3-70cb-42b8-a107-4866974c447e.jpg)

**Comment by MisterY** [comment]
I like the MagicEarth's implementation:
![[https://www.magicearth.com/wp-content/uploads/2019/12/ME_Dash_Cam-iOS.png]]

**Comment by biodranik** [comment]
What does it look like 1) in the landscape? 2) in the night mode?

**Comment by MisterY** [comment]
Hi @biodranik, please see below

![Screenshot_20220814-214152_Magic_Earth](https://user-images.githubusercontent.com/462445/184552381-3b088e31-4589-459d-811a-a3395d267b4f.png)


![Screenshot_20220814-214007_Magic_Earth](https://user-images.githubusercontent.com/462445/184552387-2ff9fc90-fdba-439a-956a-ca2e8b88d65b.png)

**Comment by biodranik** [comment]
Sorry, but that's way worse than Apple or Google )

**Comment by MisterY** [comment]
> Sorry, but that's way worse than Apple or Google )

If I look through the windshield, the exit shield is up and the lines in the lanes are down. That's exactly how they are displayed here. Simple pattern matching. The implementations with the arrows placed on top of the screen are somewhat counter-intuitive to me.

**Comment by biodranik** [comment]
1. Many people are already used to Google maps and Apple maps.
2. Lanes should be CLEARLY VISIBLE, even with bad eyesight.

**Comment by fw190d13** [comment]
Hi!
Could someone from the team explain a non tech person if implementing lane assist into OrganicMaps is difficult from technical point of view?

As I understand from the previous posts all of the data is here as shown in OsmAnd.

Lane assist is useful especially in big cities with multiple lanes and many turns and crossroads and lane changes after those crossroads.

The lack of lane assist is the only road blocker for me to switch from Magic Earth.

Not trying to be rude or push for something but just trying to understand the situation.

**Comment by biodranik** [comment]
Implementing it is not complex, as 90% of all other issues or requests. It just requires a lot of our time/resources, which are very limited at the moment :)

**Comment by ekimber** [comment]
> > Lanes are not well-mapped yet in OpenStreetMap.
> 
> According to taginfo [turn:lanes](https://taginfo.openstreetmap.org/keys/turn:lanes) tag has 683394 usages. Mapping is far from being completed but another app using this data would likely cause increase in the mapping velocity - it's quite easy to map from satellite imagery.

Yes, I occasionally find incorrect or missing lanes on OsmAnd and subsequently fix them in OSM.

The clean UI and simplicity of this app is very appealing (especially compared to OsmAnd) but lane info is definitely one of the features I feel is missing.

**Comment by Donkey-Doug** [comment]
Hi, could you maybe tell me how to fix lanes in OSM? There is a part of the highway near me incorrect in OSM, but I never managed to find out how to fix it.

**Comment by pm4rcin** [comment]
I think the ideal would be to take apple's lane assist and Magic Earth's next street names (destination) signs (with road shields of course).

**Comment by Donkey-Doug** [comment]
Thanks everybody!

**Comment by biodranik** [comment]
Reopening it, because it is not done on iOS yet, and a better UI is needed.

**Comment by oleg-rswll** [comment]
2/9/25
**Google Play Store**
Review

> Exactly what I was looking for! Solid display of towns and terrain on the map, and very decent navigation. It calculates really good routes! But I have noticed a few issues and suggestions for improvement:
> 
> - When lane navigation appears, it does not display the correct combined arrow for lanes that allow both straight and turning.
> - It says "exit" instead of "take the off-ramp."
> - I can disable all toll roads, but in Europe, this is not ideal because vignettes are purchased separately for each country, and sometimes it is not worth buying one.
